src/main/groovy/lang/ObjectRange.java

```diff
@@ -409,106 +406,86 @@ public void step(int step, Closure closure) {
             if (desiredStep == 0 && compareNotEqual(range.getFrom(), range.getTo())) {
                 throw new GroovyRuntimeException("Infinite loop detected due to step size of 0");
             }
+
             this.range = range;
-            if (range.isReverse()) {
-                step = -desiredStep;
-            } else {
-                step = desiredStep;
-            }
-            if (step > 0) {
-                value = range.getFrom();
+            if (desiredStep < 0) {
+                isAscending = range.isReverse();
             } else {
-                value = range.getTo();
+                isAscending = !range.isReverse();
             }
         }
 
         @Override
-        public void remove() {
-            range.remove(index);
+        public boolean hasNext() {
+            fetchNextIfNeeded();
+            return (next != null) && (isAscending ? compareLessThanEqual(next, range.getTo())
+                                                  : compareGreaterThanEqual(next, range.getFrom()));
         }
 
         @Override
         public Comparable next() {
-            // not thread safe
-            if (!nextFetched) {
-                value = peek();
-                nextFetched = true;
+            if (!hasNext()) {
+                return null;
             }
-            nextFetched = false;
-            index++;
-            return value;
-        }
 
-        @Override
-        public boolean hasNext() {
-            // not thread safe
-            if (!nextFetched) {
-                value = peek();
-                nextFetched = true;
-            }
-            return value != null;
+            fetchNextIfNeeded();
+            isNextFetched = false;
+            return next;
         }
 
-        private Comparable peek() {
-            if (step > 0) {
-                Comparable peekValue = value;
-                for (int i = 0; i < step; i++) {
-                    peekValue = (Comparable) range.increment(peekValue);
-                    // handle back to beginning due to modulo incrementing
-                    if (compareLessThanEqual(peekValue, range.from)) return null;
-                }
-                if (compareLessThanEqual(peekValue, range.to)) {
-                    return peekValue;
-                }
-            } else {
-                final int positiveStep = -step;
-                Comparable peekValue = value;
-                for (int i = 0; i < positiveStep; i++) {
-                    peekValue = (Comparable) range.decrement(peekValue);
-                    // handle back to beginning due to modulo decrementing
-                    if (compareGreaterThanEqual(peekValue, range.to)) return null;
-                }
-                if (compareGreaterThanEqual(peekValue, range.from)) {
-                    return peekValue;
+        private void fetchNextIfNeeded() {
+            if (!isNextFetched) {
+                isNextFetched = true;
+
+                if (next == null) {
```

Comment:
using a check with next = null might not be good enough. What happens if null is used as lower bound? I would like to avoid this leading to endless lopps